In [2]:
# Import required libraries and dependencies
import pandas as pd
import numpy as np
from pathlib import Path
import hvplot.pandas

from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

from sklearn.metrics import confusion_matrix, classification_report
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

In [3]:
### import and read csv file

path = 'Output/full_indicators_by_country.csv'
full_indicators = pd.read_csv(path)
# full_indicators = full_indicators.rename(columns={'total':'Immigration_flow'})

full_indicators.head(5)

index         country iso3Code  clean fuels for cooking (% of population)  \
0      0     Afghanistan      AFG                                  32.400000   
1      1         Albania      ALB                                  81.455556   
2      2         Algeria      DZA                                  99.677778   
3      3  American Samoa      ASM                                   0.000000   
4      4         Andorra      AND                                 100.000000   

   Access to electricity (% of population)  Agricultural land (% )  \
0                                91.555556               58.432353   
1                                99.977778               42.382186   
2                                99.655556               17.353871   
3                                 0.000000               14.150000   
4                               100.000000               39.983452   

   Armed forces (% labor force)  precipitation (mm per year)  \
0                      2.681272                        327.0   
1                      0.614119                       1485.0   
2                      2.628728                         89.0   
3                      0.000000                          0.0   
4                      0.000000                          0.0   

   Birth rate, crude (per 1,000 people)  \
0                             36.628111   
1                             10.610444   
2                             23.088556   
3                             11.422222   
4                              6.466667   

   Central government debt, total (% of GDP)  ...  \
0                                   0.000000  ...   
1                                  78.238375  ...   
2                                   0.000000  ...   
3                                   0.000000  ...   
4                                   0.000000  ...   

   Total alcohol consumption per capita (liters of pure alcohol, projected estimates, 15+ years of age)  \
0                                           0.012778                                                      
1                                           4.822222                                                      
2                                           0.574444                                                      
3                                           0.000000                                                      
4                                          10.430000                                                      

   Total debt service (% of GNI)  \
0                       0.270084   
1                       5.576404   
2                       0.162233   
3                       0.000000   
4                       0.000000   

   Total debt service (% of exports of goods, services and primary income)  \
0                                           3.220058                         
1                                          17.127848                         
2                                           0.736472                         
3                                           0.000000                         
4                                           0.000000                         

   Unemployment advanced education  Unemployment basic education   \
0                        10.011889                       7.579333   
1                        13.158667                      10.648444   
2                        13.805556                       8.501111   
3                         0.000000                       0.000000   
4                         0.000000                       0.000000   

   Unemployment intermediate education  \
0                             9.741222   
1                            14.884000   
2                             8.757778   
3                             0.000000   
4                             0.000000   

   Unemployment, total (% of total labor force) (modeled ILO estimate)  \
0                                          11.616111               

In [4]:
# Generate summary statistics
full_indicators.describe()

index  clean fuels for cooking (% of population)  \
count  203.000000                                 203.000000   
mean   101.000000                                  61.681253   
std     58.745213                                  40.911296   
min      0.000000                                   0.000000   
25%     50.500000                                  11.344444   
50%    101.000000                                  82.744444   
75%    151.500000                                 100.000000   
max    202.000000                                 100.000000   

       Access to electricity (% of population)  Agricultural land (% )  \
count                               203.000000              203.000000   
mean                                 84.540394               36.676581   
std                                  26.359982               22.669960   
min                                   0.000000                0.000000   
25%                                  82.050000               18.275213   
50%                                  99.744444               38.527215   
75%                                 100.000000               52.313291   
max                                 100.000000               81.080429   

       Armed forces (% labor force)  precipitation (mm per year)  \
count                    203.000000                   203.000000   
mean                       1.040230                  1037.241872   
std                        1.502681                   845.406017   
min                        0.000000                     0.000000   
25%                        0.206179                   324.500000   
50%                        0.583351                   847.000000   
75%                        1.279990                  1613.000000   
max                       13.679517                  3240.000000   

       Birth rate, crude (per 1,000 people)  \
count                            203.000000   
mean                              19.211691   
std                               10.010777   
min                                6.222222   
25%                               10.493333   
50%                               16.022667   
75%                               27.434056   
max                               45.817111   

       Central government debt, total (% of GDP)  \
count                                 203.000000   
mean                                   22.630020   
std                                    38.152011   
min                                     0.000000   
25%                                     0.000000   
50%                                     0.000000   
75%                                    40.770094   
max                                   209.829585   

       Consumer price index (2010 = 100)  Control of Corruption: Estimate  \
count                         203.000000                       203.000000   
mean                          292.324973                        -0.052812   
std                          1219.684353                         0.985466   
min                             0.000000                        -1.819292   
25%                           113.581446                        -0.822233   
50%                           124.388549                        -0.232470   
75%                           159.647858                         0.577856   
max                         12920.702312                         2.245200   

       ...  \
count  ...   
mean   ...   
std    ...   
min    ...   
25%    ...   
50%    ...   
75%    ...   
max    ...   

       Total alcohol consumption per capita (liters of pure alcohol, projected estimates, 15+ years of age)  \
count                                         203.000000                                                      
mean                                            4.753558                                                      
std                                             4.069545                                                    

In [6]:
# # Plot your data to see what's in your DataFrame
# full_indicators.hvplot.line(
#     width=800,
#     height=400,
#     rot=90
# )

---

### Prepare the Data

In [7]:
full_indicators.columns

Index(['index', 'country', 'iso3Code',
       'clean fuels for cooking (% of population)',
       'Access to electricity (% of population)', 'Agricultural land (% )',
       'Armed forces (% labor force)', 'precipitation (mm per year)',
       'Birth rate, crude (per 1,000 people)',
       'Central government debt, total (% of GDP)',
       'Consumer price index (2010 = 100)', 'Control of Corruption: Estimate',
       'Cost of business start-up (% of GNI per capita)',
       'Current account balance (% of GDP)',
       'Current health expenditure (% of GDP)',
       'Death rate, crude (per 1,000 people)',
       'Debt service (PPG and IMF only, % of exports of goods, services and primary income)',
       'doing business score', 'Fertility rate, total (births per woman)',
       '% Firms using banks', 'GDP growth (annual %)',
       'GDP per capita (constant 2015 US$)',
       'GDP per capita, PPP (constant 2021 international $)',
       'GNI per capita (2015 US$)', 'government  consump

In [22]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file

indicators_scaled = StandardScaler().fit_transform(full_indicators[[

                                                                      'clean fuels for cooking (% of population)',
                                                                      'Access to electricity (% of population)', 'Agricultural land (% )',
                                                                      'Armed forces (% labor force)', 'precipitation (mm per year)',
                                                                      'Birth rate, crude (per 1,000 people)',
                                                                      'Central government debt, total (% of GDP)',
                                                                      'Consumer price index (2010 = 100)', 'Control of Corruption: Estimate',
                                                                      'Cost of business start-up (% of GNI per capita)',
                                                                      'Current account balance (% of GDP)',
                                                                      'Current health expenditure (% of GDP)',
                                                                      'Death rate, crude (per 1,000 people)',
                                                                      'Debt service (PPG and IMF only, % of exports of goods, services and primary income)',
                                                                      'doing business score', 'Fertility rate, total (births per woman)',
                                                                      '% Firms using banks', 'GDP growth (annual %)',
                                                                      'GDP per capita (constant 2015 US$)',
                                                                      'GDP per capita, PPP (constant 2021 international $)',
                                                                      'GNI per capita (2015 US$)', 'government  consumption exp (% of GDP)',
                                                                      'Government expenditure on education, total (% of GDP)',
                                                                      'Human Capital Index',
                                                                      'Industry (including construction), value added (% of GDP)',
                                                                      'Inflation, consumer prices (annual %)', 'Level of water stress',
                                                                      'Life expectancy at birth, total (years)',
                                                                      'Military expenditure (% of GDP)', 'Population density',
                                                                      'Population growth (annual %)', '% of total population below 5 meters',
                                                                      'Population living in slums (% of urban population)',
                                                                      'Population, total', 'Poverty $2.15 a day (% of population)',
                                                                      'Poverty $3.65 a day (% of population)',
                                                                      'Poverty $6.85 a day (% of population)', '%  Poverty headcount ratio ',
                                                                      'Time to start a business (days)',
                                                                      'Total alcohol consumption per capita (liters of pure alcohol, projected estimates, 15+ years of age)',
                                                                      'Total debt service (% of GNI)',
                                                                      'Total debt service (% of exports of goods, services and primary income)',
                                                                      'Unemployment advanced education', 'Unemployment basic education ',
                                                                      'Unemployment intermediate education',
                                                                      'Unemployment, total (% of total labor force) (modeled ILO estimate)',
                                                                      'Corruption index 2023', 'Global_Peace_Index_2023', 'immigration_100k'
    

                                                                      
                                                                      ]])

In [23]:
# Create a DataFrame with the scaled data
indicators_scaled_df = pd.DataFrame(indicators_scaled, columns=[
                                                               
                                                               'clean fuels for cooking (% of population)',
                                                               'Access to electricity (% of population)', 'Agricultural land (% )',
                                                               'Armed forces (% labor force)', 'precipitation (mm per year)',
                                                               'Birth rate, crude (per 1,000 people)',
                                                               'Central government debt, total (% of GDP)',
                                                               'Consumer price index (2010 = 100)', 'Control of Corruption: Estimate',
                                                               'Cost of business start-up (% of GNI per capita)',
                                                               'Current account balance (% of GDP)',
                                                               'Current health expenditure (% of GDP)',
                                                               'Death rate, crude (per 1,000 people)',
                                                               'Debt service (PPG and IMF only, % of exports of goods, services and primary income)',
                                                               'doing business score', 'Fertility rate, total (births per woman)',
                                                               '% Firms using banks', 'GDP growth (annual %)',
                                                               'GDP per capita (constant 2015 US$)',
                                                               'GDP per capita, PPP (constant 2021 international $)',
                                                               'GNI per capita (2015 US$)', 'government  consumption exp (% of GDP)',
                                                               'Government expenditure on education, total (% of GDP)',
                                                               'Human Capital Index',
                                                               'Industry (including construction), value added (% of GDP)',
                                                               'Inflation, consumer prices (annual %)', 'Level of water stress',
                                                               'Life expectancy at birth, total (years)',
                                                               'Military expenditure (% of GDP)', 'Population density',
                                                               'Population growth (annual %)', '% of total population below 5 meters',
                                                               'Population living in slums (% of urban population)',
                                                               'Population, total', 'Poverty $2.15 a day (% of population)',
                                                               'Poverty $3.65 a day (% of population)',
                                                               'Poverty $6.85 a day (% of population)', '%  Poverty headcount ratio ',
                                                               'Time to start a business (days)',
                                                               'Total alcohol consumption per capita (liters of pure alcohol, projected estimates, 15+ years of age)',
                                                               'Total debt service (% of GNI)',
                                                               'Total debt service (% of exports of goods, services and primary income)',
                                                               'Unemployment advanced education', 'Unemployment basic education ',
                                                               'Unemployment intermediate education',
                                                               'Unemployment, total (% of total labor force) (modeled ILO estimate)',
                                                               'Corruption index 2023', 'Global_Peace_Index_2023', 'immigration_100k'
                                                        
                                                               ])

# # Copy the crypto names from the original DataFrame
indicators_scaled_df['country'] = full_indicators['country']
# indicators_scaled_df['iso3Code'] = full_indicators['iso3Code']



# Display the scaled DataFrame
indicators_scaled_df.head()

clean fuels for cooking (% of population)  \
0                                  -0.717495   
1                                   0.484541   
2                                   0.931050   
3                                  -1.511410   
4                                   0.938946   

   Access to electricity (% of population)  Agricultural land (% )  \
0                                 0.266787                0.962047   
1                                 0.587085                0.252303   
2                                 0.574831               -0.854456   
3                                -3.215078               -0.996132   
4                                 0.587930                0.146231   

   Armed forces (% labor force)  precipitation (mm per year)  \
0                      1.094776                    -0.842196   
1                     -0.284268                     0.530946   
2                      1.059723                    -1.124414   
3                     -0.693961                    -1.229949   
4                     -0.693961                    -1.229949   

   Birth rate, crude (per 1,000 people)  \
0                              1.744068   
1                             -0.861323   
2                              0.388226   
3                             -0.780032   
4                             -1.276278   

   Central government debt, total (% of GDP)  \
0                                  -0.594620   
1                                   1.461151   
2                                  -0.594620   
3                                  -0.594620   
4                                  -0.594620   

   Consumer price index (2010 = 100)  Control of Corruption: Estimate  \
0                          -0.120339                        -1.340081   
1                          -0.140730                        -0.454941   
2                          -0.111938                        -0.597012   
3                          -0.240265                         1.489307   
4                          -0.240265                         1.286279   

   Cost of business start-up (% of GNI per capita)  ...  \
0                                        -0.071295  ...   
1                                        -0.264877  ...   
2                                        -0.256378  ...   
3                                        -0.579959  ...   
4                                        -0.579959  ...   

   Total debt service (% of GNI)  \
0                      -0.574455   
1                       0.467228   
2                      -0.595627   
3                      -0.627475   
4                      -0.627475   

   Total debt service (% of exports of goods, services and primary income)  \
0                                          -0.492501                         
1                                           0.646547                         
2                                          -0.695907                         
3                                          -0.756224                         
4                                          -0.756224                         

   Unemployment advanced education  Unemployment basic education   \
0                         0.992769                       0.136667   
1                         1.641097                       0.587503   
2                         1.774375                       0.272071   
3                        -1.069970                      -0.976697   
4                        -1.069970                      -0.976697   

   Unemployment intermediate education  \
0                             0.592845   
1                             1.528103   
2                             0.413997   
3                            -1.178681   
4                            -1.178681   

   Unemployment, total (% of total labor force) (modeled ILO estimate)  \
0                                           0.868266                     
1                                           1.042535            

In [24]:
## Rename the indicators to be more ... friendly

indicators_scaled_df = indicators_scaled_df.rename(columns={


                                            'Access to clean fuels and technologies for cooking (% of population)'    :   'clean fuels for cooking (% of population)'  ,
                                            'Access to electricity (% of population)'    :   'Access to electricity (% of population)'  ,
                                            'Agricultural land (% of land area)'    :   'Agricultural land (% )'  ,
                                            'Armed forces personnel (% of total labor force)'    :   'Armed forces (% labor force)'  ,
                                            'Average precipitation in depth (mm per year)'    :   'precipitation (mm per year)'  ,
                                            'Birth rate crude (per 1000 people)'    :   'Birth rate crude'  ,
                                            'Central government debt total (% of GDP)'    :   'government debt (% of GDP)'  ,
                                            'Consumer price index (2010 = 100) Control of Corruption: Estimate'    :   'Consumer price index (2010)'  ,
                                            'Cost of business start-up procedures (% of GNI per capita)'    :   'Cost of business start-up (% of GNI per capita)'  ,
                                            'Current account balance (% of GDP)'    :   'Current account balance (% of GDP)'  ,
                                            'Current health expenditure (% of GDP)'    :   'Current health expenditure (% of GDP)'  ,
                                            'Death rate crude (per 1000 people)'    :   'Death rate crude (per 1000 people)'  ,
                                            'Debt service (PPG and IMF only % of exports of goods services and primary income)'    :   'Debt service'  ,
                                            'Ease of doing business score (0 = lowest performance to 100 = best performance)'    :   'doing business score'  ,
                                            'Fertility rate total (births per woman)'    :   'Fertility rate per woman'  ,
                                            'Firms using banks to finance investment (% of firms)'    :   '% Firms using banks'  ,
                                            'GDP growth (annual %) GDP per capita (constant 2015 US$)'    :   'GDP per capita growth '  ,
                                            'GDP per capita PPP (constant 2021 international $)'    :   'GDP per capita PPP ( 2021)'  ,
                                            'GNI per capita (constant 2015 US$)'    :   'GNI per capita (2015 US$)'  ,
                                            'General government final consumption expenditure (% of GDP)'    :   'government  consumption exp (% of GDP)'  ,
                                            'Government expenditure on education total (% of GDP)'    :   'Government expenditure on education total (% of GDP)'  ,
                                            'Human Capital Index (HCI) (scale 0-1)'    :   'Human Capital Index'  ,
                                            'Industry (including construction) value added (% of GDP)'    :   'Industry (% of GDP)'  ,
                                            'Inflation consumer prices (annual %)'    :   'Inflation consumer prices (%)'  ,
                                            'Level of water stress: freshwater withdrawal as a proportion of available freshwater resources'    :   'Level of water stress'  ,
                                            'Life expectancy at birth total (years)'    :   'Life expectancy at birth total (years)'  ,
                                            'Military expenditure (% of GDP)'    :   'Military expenditure (% of GDP)'  ,
                                            'Population density (people per sq. km of land area)'    :   'Population density'  ,
                                            'Population growth (annual %)'    :   'Population growth (annual %)'  ,
                                            'Population living in areas where elevation is below 5 meters (% of total population)'    :   '% of total population below 5 meters'  ,
                                            'Population living in slums (% of urban population)'    :   'Population living in slums (% of urban population)'  ,
                                            'Population total'    :   'Population total'  ,
                                            'Poverty headcount ratio at $2.15 a day (2017 PPP) (% of population)'    :   'Poverty $2.15 a day (% of population)'  ,
                                            'Poverty headcount ratio at $3.65 a day (2017 PPP) (% of population)'    :   'Poverty $3.65 a day (% of population)'  ,
                                            'Poverty headcount ratio at $6.85 a day (2017 PPP) (% of population)'    :   'Poverty $6.85 a day (% of population)'  ,
                                            'Poverty headcount ratio at national poverty lines (% of population)'    :   '%  Poverty headcount ratio '  ,
                                            'Time required to start a business (days)'    :   'Time to start a business (days)'  ,
                                            'Total alcohol consumption per capita (liters of pure alcohol projected estimates 15+ years of age)'    :   'Alcohol consumption per capita'  ,
                                            'Total debt service (% of GNI)'    :   'Total debt service (% of GNI)'  ,
                                            'Total debt service (% of exports of goods services and primary income)'    :   'Total debt service (% of exports)'  ,
                                            'Unemployment with advanced education (% of total labor force with advanced education)'    :   'Unemployment advanced education'  ,
                                            'Unemployment with basic education (% of total labor force with basic education)'    :   'Unemployment basic education '  ,
                                            'Unemployment with intermediate education (% of total labor force with intermediate education)'    :   'Unemployment intermediate education'  ,
                                            'Unemployment total (% of total labor force) (modeled ILO estimate)]'    :   'Unemployment total'  ,
                                            'CPI_score_2023' : 'Corruption index 2023'


                                                    })
indicators_scaled_df.head(5)

clean fuels for cooking (% of population)  \
0                                  -0.717495   
1                                   0.484541   
2                                   0.931050   
3                                  -1.511410   
4                                   0.938946   

   Access to electricity (% of population)  Agricultural land (% )  \
0                                 0.266787                0.962047   
1                                 0.587085                0.252303   
2                                 0.574831               -0.854456   
3                                -3.215078               -0.996132   
4                                 0.587930                0.146231   

   Armed forces (% labor force)  precipitation (mm per year)  \
0                      1.094776                    -0.842196   
1                     -0.284268                     0.530946   
2                      1.059723                    -1.124414   
3                     -0.693961                    -1.229949   
4                     -0.693961                    -1.229949   

   Birth rate, crude (per 1,000 people)  \
0                              1.744068   
1                             -0.861323   
2                              0.388226   
3                             -0.780032   
4                             -1.276278   

   Central government debt, total (% of GDP)  \
0                                  -0.594620   
1                                   1.461151   
2                                  -0.594620   
3                                  -0.594620   
4                                  -0.594620   

   Consumer price index (2010 = 100)  Control of Corruption: Estimate  \
0                          -0.120339                        -1.340081   
1                          -0.140730                        -0.454941   
2                          -0.111938                        -0.597012   
3                          -0.240265                         1.489307   
4                          -0.240265                         1.286279   

   Cost of business start-up (% of GNI per capita)  ...  \
0                                        -0.071295  ...   
1                                        -0.264877  ...   
2                                        -0.256378  ...   
3                                        -0.579959  ...   
4                                        -0.579959  ...   

   Total debt service (% of GNI)  \
0                      -0.574455   
1                       0.467228   
2                      -0.595627   
3                      -0.627475   
4                      -0.627475   

   Total debt service (% of exports of goods, services and primary income)  \
0                                          -0.492501                         
1                                           0.646547                         
2                                          -0.695907                         
3                                          -0.756224                         
4                                          -0.756224                         

   Unemployment advanced education  Unemployment basic education   \
0                         0.992769                       0.136667   
1                         1.641097                       0.587503   
2                         1.774375                       0.272071   
3                        -1.069970                      -0.976697   
4                        -1.069970                      -0.976697   

   Unemployment intermediate education  \
0                             0.592845   
1                             1.528103   
2                             0.413997   
3                            -1.178681   
4                            -1.178681   

   Unemployment, total (% of total labor force) (modeled ILO estimate)  \
0                                           0.868266                     
1                                           1.042535            

In [25]:
indicators_scaled_df.columns

Index(['clean fuels for cooking (% of population)',
       'Access to electricity (% of population)', 'Agricultural land (% )',
       'Armed forces (% labor force)', 'precipitation (mm per year)',
       'Birth rate, crude (per 1,000 people)',
       'Central government debt, total (% of GDP)',
       'Consumer price index (2010 = 100)', 'Control of Corruption: Estimate',
       'Cost of business start-up (% of GNI per capita)',
       'Current account balance (% of GDP)',
       'Current health expenditure (% of GDP)',
       'Death rate, crude (per 1,000 people)',
       'Debt service (PPG and IMF only, % of exports of goods, services and primary income)',
       'doing business score', 'Fertility rate, total (births per woman)',
       '% Firms using banks', 'GDP growth (annual %)',
       'GDP per capita (constant 2015 US$)',
       'GDP per capita, PPP (constant 2021 international $)',
       'GNI per capita (2015 US$)', 'government  consumption exp (% of GDP)',
       'Governmen

In [27]:
# Organize columns

indicators_scaled_df = indicators_scaled_df[[

                                                'country', 
                                                'clean fuels for cooking (% of population)',
                                                'Access to electricity (% of population)', 'Agricultural land (% )',
                                            'Armed forces (% labor force)', 'precipitation (mm per year)',
                                            'Birth rate, crude (per 1,000 people)',
                                            'Central government debt, total (% of GDP)',
                                            'Consumer price index (2010 = 100)', 'Control of Corruption: Estimate',
                                            'Cost of business start-up (% of GNI per capita)',
                                            'Current account balance (% of GDP)',
                                            'Current health expenditure (% of GDP)',
                                            'Death rate, crude (per 1,000 people)',
                                            'Debt service (PPG and IMF only, % of exports of goods, services and primary income)',
                                            'doing business score', 'Fertility rate, total (births per woman)',
                                            '% Firms using banks', 'GDP growth (annual %)',
                                            'GDP per capita (constant 2015 US$)',
                                            'GDP per capita, PPP (constant 2021 international $)',
                                            'GNI per capita (2015 US$)', 'government  consumption exp (% of GDP)',
                                            'Government expenditure on education, total (% of GDP)',
                                            'Human Capital Index',
                                            'Industry (including construction), value added (% of GDP)',
                                            'Inflation, consumer prices (annual %)', 'Level of water stress',
                                            'Life expectancy at birth, total (years)',
                                            'Military expenditure (% of GDP)', 'Population density',
                                            'Population growth (annual %)', '% of total population below 5 meters',
                                            'Population living in slums (% of urban population)',
                                            'Population, total', 'Poverty $2.15 a day (% of population)',
                                            'Poverty $3.65 a day (% of population)',
                                            'Poverty $6.85 a day (% of population)', '%  Poverty headcount ratio ',
                                            'Time to start a business (days)',
                                            'Total alcohol consumption per capita (liters of pure alcohol, projected estimates, 15+ years of age)',
                                            'Total debt service (% of GNI)',
                                            'Total debt service (% of exports of goods, services and primary income)',
                                            'Unemployment advanced education', 'Unemployment basic education ',
                                            'Unemployment intermediate education',
                                            'Unemployment, total (% of total labor force) (modeled ILO estimate)',
                                            'Corruption index 2023', 'Global_Peace_Index_2023', 'immigration_100k'
                                                                
                                            ]]
indicators_scaled_df.head(5)

country  clean fuels for cooking (% of population)  \
0     Afghanistan                                  -0.717495   
1         Albania                                   0.484541   
2         Algeria                                   0.931050   
3  American Samoa                                  -1.511410   
4         Andorra                                   0.938946   

   Access to electricity (% of population)  Agricultural land (% )  \
0                                 0.266787                0.962047   
1                                 0.587085                0.252303   
2                                 0.574831               -0.854456   
3                                -3.215078               -0.996132   
4                                 0.587930                0.146231   

   Armed forces (% labor force)  precipitation (mm per year)  \
0                      1.094776                    -0.842196   
1                     -0.284268                     0.530946   
2                      1.059723                    -1.124414   
3                     -0.693961                    -1.229949   
4                     -0.693961                    -1.229949   

   Birth rate, crude (per 1,000 people)  \
0                              1.744068   
1                             -0.861323   
2                              0.388226   
3                             -0.780032   
4                             -1.276278   

   Central government debt, total (% of GDP)  \
0                                  -0.594620   
1                                   1.461151   
2                                  -0.594620   
3                                  -0.594620   
4                                  -0.594620   

   Consumer price index (2010 = 100)  Control of Corruption: Estimate  ...  \
0                          -0.120339                        -1.340081  ...   
1                          -0.140730                        -0.454941  ...   
2                          -0.111938                        -0.597012  ...   
3                          -0.240265                         1.489307  ...   
4                          -0.240265                         1.286279  ...   

   Total alcohol consumption per capita (liters of pure alcohol, projected estimates, 15+ years of age)  \
0                                          -1.167821                                                      
1                                           0.016914                                                      
2                                          -1.029463                                                      
3                                          -1.170969                                                      
4                                           1.398308                                                      

   Total debt service (% of GNI)  \
0                      -0.574455   
1                       0.467228   
2                      -0.595627   
3                      -0.627475   
4                      -0.627475   

   Total debt service (% of exports of goods, services and primary income)  \
0                                          -0.492501                         
1                                           0.646547                         
2                                          -0.695907                         
3                                          -0.756224                         
4                                          -0.756224                         

   Unemployment advanced education  Unemployment basic education   \
0                         0.992769                       0.136667   
1                         1.641097                       0.587503   
2                         1.774375                       0.272071   
3                        -1.069970                      -0.976697   
4                        -1.069970                      -0.976697   

   Unemployment intermediate education  \
0               

In [28]:
indicators_scaled_df = indicators_scaled_df.set_index(['country'])
indicators_scaled_df.head(5)

clean fuels for cooking (% of population)  \
country                                                     
Afghanistan                                     -0.717495   
Albania                                          0.484541   
Algeria                                          0.931050   
American Samoa                                  -1.511410   
Andorra                                          0.938946   

                Access to electricity (% of population)  \
country                                                   
Afghanistan                                    0.266787   
Albania                                        0.587085   
Algeria                                        0.574831   
American Samoa                                -3.215078   
Andorra                                        0.587930   

                Agricultural land (% )  Armed forces (% labor force)  \
country                                                                
Afghanistan                   0.962047                      1.094776   
Albania                       0.252303                     -0.284268   
Algeria                      -0.854456                      1.059723   
American Samoa               -0.996132                     -0.693961   
Andorra                       0.146231                     -0.693961   

                precipitation (mm per year)  \
country                                       
Afghanistan                       -0.842196   
Albania                            0.530946   
Algeria                           -1.124414   
American Samoa                    -1.229949   
Andorra                           -1.229949   

                Birth rate, crude (per 1,000 people)  \
country                                                
Afghanistan                                 1.744068   
Albania                                    -0.861323   
Algeria                                     0.388226   
American Samoa                             -0.780032   
Andorra                                    -1.276278   

                Central government debt, total (% of GDP)  \
country                                                     
Afghanistan                                     -0.594620   
Albania                                          1.461151   
Algeria                                         -0.594620   
American Samoa                                  -0.594620   
Andorra                                         -0.594620   

                Consumer price index (2010 = 100)  \
country                                             
Afghanistan                             -0.120339   
Albania                                 -0.140730   
Algeria                                 -0.111938   
American Samoa                          -0.240265   
Andorra                                 -0.240265   

                Control of Corruption: Estimate  \
country                                           
Afghanistan                           -1.340081   
Albania                               -0.454941   
Algeria                               -0.597012   
American Samoa                         1.489307   
Andorra                                1.286279   

                Cost of business start-up (% of GNI per capita)  ...  \
country                                                          ...   
Afghanistan                                           -0.071295  ...   
Albania                                               -0.264877  ...   
Algeria                                               -0.256378  ...   
American Samoa                                        -0.579959  ...   
Andorra                                               -0.579959  ...   

                Total alcohol consumption per capita (liters of pure alcohol, projected estimates, 15+ years of age)  \
country                                                                                                                
Afghanistan                                

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [29]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,12))
k

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [30]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model = KMeans(n_clusters=i, random_state=1)
    k_model.fit(indicators_scaled_df)
    inertia.append(k_model.inertia_)


c:\Users\nazim\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\nazim\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\nazim\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\nazim\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Window

In [31]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)
df_elbow.head()

k      inertia
0  1  9947.000000
1  2  8537.296289
2  3  7853.474127
3  4  7179.782830
4  5  6921.395787

In [32]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_original_data = df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve without PCA", 
    xticks=k
)

elbow_original_data


:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:**  4

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [33]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=0)

In [34]:
# Fit the K-Means model using the scaled DataFrame
model.fit(indicators_scaled_df)

c:\Users\nazim\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=0)

In [35]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
kmeans_predictions = model.predict(indicators_scaled_df)

# Print the resulting array of cluster values.
print(kmeans_predictions)

[2 3 3 0 0 1 0 3 3 0 0 3 0 0 1 0 3 0 3 1 0 3 3 3 3 3 0 0 0 1 1 3 2 2 0 2 1
 0 0 3 2 1 2 3 1 0 0 0 0 0 0 3 0 3 3 3 3 2 2 0 1 0 0 0 0 2 1 3 0 1 3 0 3 1
 1 0 2 3 0 0 0 1 3 3 2 0 0 0 3 0 3 3 1 0 2 0 3 0 3 2 0 3 1 1 2 0 0 0 0 2 1
 0 0 1 0 0 1 3 3 0 0 3 3 3 1 2 0 3 0 0 0 3 1 1 3 0 0 0 1 0 0 2 3 3 3 0 0 0
 0 0 0 1 3 0 1 0 1 3 0 1 0 0 0 0 0 2 3 2 0 3 3 0 2 3 0 0 2 3 1 0 0 1 0 0 3
 3 2 0 0 1 3 0 0 0 0 3 0 2 0 3 2 1 2]


In [39]:
# Create a copy of the scaled DataFrame
indicators_prediction_df = indicators_scaled_df.copy()

In [40]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
indicators_prediction_df['countries_cluster_original_data'] = kmeans_predictions

# Display the copy of the scaled DataFrame
indicators_prediction_df.head()

clean fuels for cooking (% of population)  \
country                                                     
Afghanistan                                     -0.717495   
Albania                                          0.484541   
Algeria                                          0.931050   
American Samoa                                  -1.511410   
Andorra                                          0.938946   

                Access to electricity (% of population)  \
country                                                   
Afghanistan                                    0.266787   
Albania                                        0.587085   
Algeria                                        0.574831   
American Samoa                                -3.215078   
Andorra                                        0.587930   

                Agricultural land (% )  Armed forces (% labor force)  \
country                                                                
Afghanistan                   0.962047                      1.094776   
Albania                       0.252303                     -0.284268   
Algeria                      -0.854456                      1.059723   
American Samoa               -0.996132                     -0.693961   
Andorra                       0.146231                     -0.693961   

                precipitation (mm per year)  \
country                                       
Afghanistan                       -0.842196   
Albania                            0.530946   
Algeria                           -1.124414   
American Samoa                    -1.229949   
Andorra                           -1.229949   

                Birth rate, crude (per 1,000 people)  \
country                                                
Afghanistan                                 1.744068   
Albania                                    -0.861323   
Algeria                                     0.388226   
American Samoa                             -0.780032   
Andorra                                    -1.276278   

                Central government debt, total (% of GDP)  \
country                                                     
Afghanistan                                     -0.594620   
Albania                                          1.461151   
Algeria                                         -0.594620   
American Samoa                                  -0.594620   
Andorra                                         -0.594620   

                Consumer price index (2010 = 100)  \
country                                             
Afghanistan                             -0.120339   
Albania                                 -0.140730   
Algeria                                 -0.111938   
American Samoa                          -0.240265   
Andorra                                 -0.240265   

                Control of Corruption: Estimate  \
country                                           
Afghanistan                           -1.340081   
Albania                               -0.454941   
Algeria                               -0.597012   
American Samoa                         1.489307   
Andorra                                1.286279   

                Cost of business start-up (% of GNI per capita)  ...  \
country                                                          ...   
Afghanistan                                           -0.071295  ...   
Albania                                               -0.264877  ...   
Algeria                                               -0.256378  ...   
American Samoa                                        -0.579959  ...   
Andorra                                               -0.579959  ...   

                Total debt service (% of GNI)  \
country                                         
Afghanistan                         -0.574455   
Albania                              0.467228   
Algeria                             -0.595627   
American Samoa                      -0

In [44]:
# Create a scatter plot using hvPlot by setting
    # `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
    # Color the graph points with the labels found using K-Means and
    # add the crypto name in the `hover_cols` parameter to identify
    # the cryptocurrency represented by each data point.

clusters_original_data = indicators_prediction_df.hvplot.scatter(
    x="immigration_100k", 
    y="Global_Peace_Index_2023", 
    by="countries_cluster_original_data",
    title = "Clusters with original data",
    hover_cols = 'country'
)

clusters_original_data

:NdOverlay   [countries_cluster_original_data]
   :Scatter   [immigration_100k]   (Global_Peace_Index_2023,country)

---

### Optimize Clusters with Principal Component Analysis.

In [48]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=4)

In [49]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
indicators_pca = pca.fit_transform(indicators_scaled_df)

# View the scaled PCA data
indicators_pca[0:5]

array([[ 3.23200025, -0.8283388 ,  0.48018802,  0.33434859],
       [-0.89190045, -2.48232244,  2.23019704, -1.04857731],
       [-0.27982685, -0.47773574,  0.92085112,  2.24533521],
       [ 0.6293702 ,  6.76548966,  1.40147728, -2.73509082],
       [-2.0243087 ,  5.0460066 ,  0.21156073, -1.89265386]])

In [50]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.20445913, 0.10960015, 0.06726142, 0.05579026])

In [51]:
# Use 'sum' on the explained varaibles array to obtain the total explained variance

total_explained_variance = sum(pca.explained_variance_ratio_)
print(f'The total explained variance is : {total_explained_variance}')

The total explained variance is : 0.43711096396712923


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 43.7%

In [70]:
# Create a new DataFrame with the PCA data.
indicators_pca_df = pd.DataFrame(
    indicators_pca,
    columns=["PCA1", "PCA2", 'PCA3', 'PCA4']
)

# Copy the crypto names from the original scaled DataFrame
indicators_pca_df['country'] = full_indicators['country']

# # Set the coin_id column as index
indicators_pca_df = indicators_pca_df.set_index(['country'])

# Display the scaled PCA DataFrame
indicators_pca_df.head(5)

PCA1      PCA2      PCA3      PCA4
country                                               
Afghanistan     3.232000 -0.828339  0.480188  0.334349
Albania        -0.891900 -2.482322  2.230197 -1.048577
Algeria        -0.279827 -0.477736  0.920851  2.245335
American Samoa  0.629370  6.765490  1.401477 -2.735091
Andorra        -2.024309  5.046007  0.211561 -1.892654

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [71]:
# Create a list with the number of k-values from 1 to 11
    # we already have the list
k

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [72]:
# Create an empty list to store the inertia values

inertia_pca = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model_pca = KMeans(n_clusters=i, random_state=1)
    k_model_pca.fit(indicators_pca_df)
    inertia_pca.append(k_model_pca.inertia_)

c:\Users\nazim\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\nazim\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\nazim\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\nazim\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Window

In [73]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca = {"k": k, "inertia": inertia_pca}

# Create a DataFrame with the data to plot the Elbow curve

df_elbow_pca = pd.DataFrame(elbow_data_pca)
df_elbow_pca.head()


k      inertia
0  1  4347.942759
1  2  3010.065977
2  3  2302.609150
3  4  1959.614633
4  5  1379.045686

In [74]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.

elbow_with_pca = df_elbow_pca.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve_with_PCA", 
    xticks=k
)

elbow_with_pca

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 5


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** yes

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [75]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=0)

In [76]:
# Fit the K-Means model using the PCA data
model.fit(indicators_pca_df)

c:\Users\nazim\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=0)

In [77]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
kmeans_predictions_pca = model.predict(indicators_pca_df)

# Print the resulting array of cluster values.
print(kmeans_predictions_pca)

[1 3 3 2 2 1 2 3 3 0 0 3 2 0 1 0 3 0 3 1 2 3 3 3 3 3 2 0 3 1 1 3 2 1 0 1 1
 0 0 3 2 1 1 3 1 0 0 2 0 0 0 3 3 3 3 3 3 2 2 0 1 3 0 0 2 3 1 3 0 1 3 3 3 1
 1 3 2 1 0 0 0 1 3 3 3 0 0 0 3 0 3 3 1 2 2 0 3 0 3 3 0 3 1 1 2 2 0 0 2 1 1
 0 2 1 0 2 1 3 3 3 2 3 3 3 1 1 2 3 0 2 0 3 1 1 3 2 0 0 1 2 3 2 3 3 3 0 0 2
 0 0 0 1 2 2 1 0 1 3 0 1 0 2 0 0 2 1 3 1 0 3 3 3 1 2 0 0 2 1 1 0 2 1 2 0 3
 3 2 2 2 1 3 0 0 0 0 3 2 2 3 3 2 1 1]


In [78]:
# Create a copy of the scaled PCA DataFrame

indicators_pca_new_df = indicators_pca_df.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
indicators_pca_new_df['countries_cluster_withPCA'] = kmeans_predictions_pca

# Display the copy of the scaled PCA DataFrame
indicators_pca_new_df.head(5)

PCA1      PCA2      PCA3      PCA4  \
country                                                  
Afghanistan     3.232000 -0.828339  0.480188  0.334349   
Albania        -0.891900 -2.482322  2.230197 -1.048577   
Algeria        -0.279827 -0.477736  0.920851  2.245335   
American Samoa  0.629370  6.765490  1.401477 -2.735091   
Andorra        -2.024309  5.046007  0.211561 -1.892654   

                countries_cluster_withPCA  
country                                    
Afghanistan                             1  
Albania                                 3  
Algeria                                 3  
American Samoa                          2  
Andorra                                 2

In [79]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
clusters_with_PCA = indicators_pca_new_df.hvplot.scatter(
    x="PCA1", 
    y="PCA2", 
    by="countries_cluster_withPCA",
    title = "Clusters with PCA" ,
    hover_cols = 'country'
)

clusters_with_PCA

:NdOverlay   [countries_cluster_withPCA]
   :Scatter   [PCA1]   (PCA2,country)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [80]:
# Composite plot to contrast the Elbow curves

plots = (elbow_original_data + elbow_with_pca).cols(1)
plots

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [81]:
# Composite plot to contrast the clusters
# YOUR CODE HERE!


plots = (clusters_original_data + clusters_with_PCA).cols(1)
plots

:Layout
   .NdOverlay.I  :NdOverlay   [countries_cluster_original_data]
      :Scatter   [immigration_100k]   (Global_Peace_Index_2023,country)
   .NdOverlay.II :NdOverlay   [countries_cluster_withPCA]
      :Scatter   [PCA1]   (PCA2,country)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** 

  1)  regarding the number of clusters
  the fact that the optimal k was the same with and without use of PCA suggests the following : 
  - PCA did not significantly distort the structure of the data relevant to clustering..
  - The original dataset likely had a strong cluster structure and the clustering information was present in the original feature set.
  - the clusters in the original dataset are well separated, so the PCA did not change significantly the separability.

by another hand this consistancy is positive because 
  - it gives more confidence in the clustering results


  2) regarding the clustering assignements
  we observe that both methods (with and without PCA) give the same clusters (whatever is the # of th cluster)
  all the coins belong to the same cluster in both methods as shown in the dataframe below

  this reinforce the conclusions above

In [86]:
# to have the detailled content of the clusters, we will merge the 2 dataframes (with original data and with PCA)
# than we will compare the content of each cluster
# no need to make further queries, the data is limited and allows us to conclude

clusters = pd.merge(indicators_pca_new_df, indicators_prediction_df, on='country')
clusters = clusters[['countries_cluster_original_data', 'countries_cluster_withPCA']].sort_values(['countries_cluster_withPCA'])

# clusters.head(5)
clusters

countries_cluster_original_data  countries_cluster_withPCA
country                                                                       
Iceland                                           0                          0
Latvia                                            0                          0
Germany                                           0                          0
Russian Federation                                0                          0
Romania                                           0                          0
...                                             ...                        ...
Morocco                                           3                          3
Montenegro                                        3                          3
Mongolia                                          3                          3
Mexico                                            3                          3
Lebanon                                           3                          3

[203 rows x 2 columns]

In [87]:
clusters.to_csv('Output_analysis/Clusters_analysis.csv')